Go codegen doesn't work

**Go codegen doesn't work**
    
    $ pants export-codegen ::

    11:11:48.11 [ERROR] 1 Exception encountered:
    
    Engine traceback:
    in `export-codegen` goal
    
    ProcessExecutionFailure: Process 'Build Go gRPC protobuf plugin for `protoc`.' failed with exit code 1.
    stdout:
    
    stderr:
    go: google.golang.org/grpc/cmd/protoc-gen-go-grpc@v1.2.0: loading deprecation for google.golang.org/grpc/cmd/protoc-gen-go-grpc: module lookup disabled by GOPROXY=off


So, `export-codegen` doesn't work. Also, the [docs](https://www.pantsbuild.org/stable/docs/go/integrations/protobuf) describe that Pants auto-regenerates any protobuf output. But it doesn't. 
If you remove the `pkg/person.pb.go `file, and run

    pants test :: tests fail. If you run protoc --go_out=. --go_opt=paths=source_relative person.proto` 

(in `pkg`), tests are good again.

So, about `GOPROXY`: I've tried setting it multiple ways, but Pants seems to ignore it all. I think there may be a relation with: #21527

**Pants version**
Tested on `2.22` and `2.23`

**OS**
Both Linux and Mac.

**Additional info**
Here's a minimal reproduction repo: https://github.com/rdeknijf/pants-repro-protoc

--- a/src/python/pants/backend/codegen/protobuf/go/rules.py
+++ b/src/python/pants/backend/codegen/protobuf/go/rules.py
@@ -598,6 +598,8 @@ async def setup_go_protoc_plugin() -> _SetupGoProtocPlugin:
                 input_digest=download_sources_result.output_digest,
                 output_files=["gopath/bin/protoc-gen-go"],
                 description="Build Go protobuf plugin for `protoc`.",
+                # Allow `go` to contact the Go module proxy since it will run its own build.
+                allow_downloads=True,
             ),
         ),
         Get(
@@ -610,6 +612,8 @@ async def setup_go_protoc_plugin() -> _SetupGoProtocPlugin:
                 input_digest=download_sources_result.output_digest,
                 output_files=["gopath/bin/protoc-gen-go-grpc"],
                 description="Build Go gRPC protobuf plugin for `protoc`.",
+                # Allow `go` to contact the Go module proxy since it will run its own build.
+                allow_downloads=True,
             ),
         ),
     )

